### PR TITLE
compressDrv and compressDrvWeb follow up

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/compress-drv/default.nix [hunk 1]
```diff
@@ -51,7 +51,11 @@
   :::
 */
 drv:
-{ formats, compressors }:
+{
+  formats,
+  compressors,
+  extraArgs ? "",
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:57]: Presence of `extraArgs` means we will never be able to replace `find` with something else, as this broadens up the interface up to the point where we cannot reason how it works.

AFAIK from your comment earlier, would you like to exclude some extensions? If yes, how about `excludePatterns` or so?

File: pkgs/tools/compression/zopfli/default.nix [hunk 2]
```diff
@@ -34,6 +34,7 @@ stdenv.mkDerivation rec {
     '';
     platforms = platforms.unix;
     license = licenses.asl20;
+    mainProgram = "zopfli";
```
Comment (motiejus) [inline pkgs/tools/compression/zopfli/default.nix:37]: Let's move this and the `zstd` one to a separate PR.

**Review by motiejus — CHANGES_REQUESTED** [review]

**Comment by motiejus** [comment]
> I have read somewhere (can't find links handy) that for web server case, zstd may not be as useful as `brotli` due to longer decompression speed, but I may be wrong here.

I spent some time looking into this, it is [pure FUD](https://jakstys.lt/2024/web-compression/). I should know better than that before writing such statements. Sorry.

I will happily accept `zstd`.

File: pkgs/tools/compression/zopfli/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/tools/compression/zopfli/default.nix:37]: Doesn't really matter, we can just merge this as is.

File: pkgs/build-support/compress-drv/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:57]: I want to exclude some paths but good point.

Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:57]: that's what I am using right now

```
package = pkgs.compressDrvWeb pkgs.nextcloud29 {
  excludePatterns = [ ".*(\/apps\/.*\/l10n\/).*" ];
};
```

File: pkgs/applications/office/paperless-ngx/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/office/paperless-ngx/default.nix:1]: I am only compressing zstd files for paperless, since it already does br, gz by iteself via I think ``manage.py collectstatic`` and overwriting brotli files doesn't work because they are rrr.

**Comment by motiejus** [comment]
> > * compression level is unclear. `zopfi` and `brotli` default to max, but with `zstd` it's different -- should we use 3? 6? 19? 22? Ultra? Higher decompression levels require more memory.
> 
> We don't set any levels on the other compression tools, just stick to the default which should be good enough.

zopfli and brotli default to max compression, because they are designed for fast decompression. `zstd` is balancing on both, so it's default is very low.

Here is brotli vs `zstd -19` from my [last experiment](https://jakstys.lt/2024/web-compression/) (compression ratio is only 4% different):


| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `./zstd-19` | 10.7 ± 0.5 | 9.7 | 14.6 | 1.00 |
| `./brotli` | 15.1 ± 0.5 | 14.2 | 18.2 | 1.41 ± 0.08 |


```
Benchmark 1 (460 runs): ./zstd-19
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          10.8ms ±  465us    9.94ms … 14.0ms         19 ( 4%)        0%
  peak_rss           12.5MB ± 86.9KB    12.2MB … 12.7MB          1 ( 0%)        0%
  cpu_cycles         30.6M  ±  919K     29.7M  … 36.3M          41 ( 9%)        0%
  instructions       88.4M  ± 12.6K     88.4M  … 88.5M          11 ( 2%)        0%
  cache_references   1.80M  ± 22.2K     1.76M  … 2.01M          26 ( 6%)        0%
  cache_misses        188K  ± 3.66K      179K  …  222K           7 ( 2%)        0%
  branch_misses       346K  ± 1.02K      344K  …  352K          24 ( 5%)        0%
Benchmark 2 (328 runs): ./brotli
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          15.2ms ±  488us    14.3ms … 18.9ms         22 ( 7%)        💩+ 40.5% ±  0.6%
  peak_rss           12.0MB ± 91.4KB    11.8MB … 12.2MB          0 ( 0%)        ⚡-  3.7% ±  0.1%
  cpu_cycles         52.6M  ±  575K     51.6M  … 58.9M          11 ( 3%)        💩+ 71.7% ±  0.4%
  instructions        101M  ± 13.1K      101M  …  101M          10 ( 3%)        💩+ 14.5% ±  0.0%
  cache_references   1.95M  ± 75.4K     1.92M  … 3.25M          10 ( 3%)        💩+  8.3% ±  0.4%
  cache_misses        161K  ± 2.37K      153K  …  180K           5 ( 2%)        ⚡- 14.7% ±  0.2%
  branch_misses       898K  ±  869       895K  …  902K           8 ( 2%)        💩+159.6% ±  0.0%
```

As far as API goes, I think I may reverse the direction: the possible incarnations of `find` are unbounded, so wrapping them all can be futile. Please give me some time to think/experiment about it.

File: pkgs/build-support/compress-drv/default.nix [hunk 3]
```diff
@@ -13,6 +13,10 @@
 
   : List of file extensions to compress. Example: `["txt" "svg" "xml"]`.
 
+  `extraFindOperands` (String)
+
+  : Exclude pattern. Example: `-not -iregex ".*(\/apps\/.*\/l10n\/).*"`
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:18]: Leftover comment

File: pkgs/build-support/compress-drv/default.nix [hunk 4]
```diff
@@ -65,14 +73,22 @@ let
...
-  formatsPipe = builtins.concatStringsSep "|" formats;
+  formatsPipe = lib.concatStringsSep "|" formats;
 in
-runCommand "${drv.name}-compressed" { } ''
-  mkdir $out
-  (cd $out; ${xorg.lndir}/bin/lndir ${drv})
+runCommand "${drv.name}-compressed"
+  (
+    {
+      nativeBuildInputs = [ xorg.lndir ];
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:84]: Why this insteas of full path?

File: pkgs/build-support/compress-drv/default.nix [hunk 5]
```diff
@@ -65,14 +73,22 @@ let
...
+    }
+    // (lib.optionalAttrs (drv ? pname) { inherit (drv) pname; })
+    // (lib.optionalAttrs (drv ? version) { inherit (drv) version; })
+  )
+  ''
+    mkdir $out
+    (cd $out; lndir -silent ${drv})
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:91]: Why silent?

File: pkgs/build-support/compress-drv/default.nix [hunk 6]
```diff
@@ -65,14 +73,22 @@ let
...
-runCommand "${drv.name}-compressed" { } ''
-  mkdir $out
-  (cd $out; ${xorg.lndir}/bin/lndir ${drv})
+runCommand "${drv.name}-compressed"
+  (
+    {
+      nativeBuildInputs = [ xorg.lndir ];
+    }
+    // (lib.optionalAttrs (drv ? pname) { inherit (drv) pname; })
+    // (lib.optionalAttrs (drv ? version) { inherit (drv) version; })
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:87]: Can you help me understand why those 2 are needed for the wrappeijg derivation?

**Comment by philiptaron** [comment]
I don't understand the need for `extraFindOperands` given the current usecase.

**Comment by motiejus** [comment]
> I don't understand the need for `extraFindOperands` given the current usecase.

https://github.com/NixOS/nixpkgs/pull/239198

Tldr: it needs to exclude `vendor` directory.

File: pkgs/by-name/gi/gitea/package.nix [hunk 7]
```diff
@@ -88,7 +88,7 @@ in buildGoModule rec {
   '';
 
   passthru = {
-    data-compressed = lib.warn "gitea.passthru.data-compressed is deprecated. Use \"compressDrvWeb gitea.data\"." (compressDrvWeb gitea.data);
+    data-compressed = lib.warn "gitea.passthru.data-compressed is deprecated. Use \"compressDrvWeb gitea.data\"." (compressDrvWeb gitea.data { });
```
Comment (philiptaron) [inline pkgs/by-name/gi/gitea/package.nix:91]: When I run this on master, I get a bunch of `Permission denied` errors. This commit doesn't change that outcome.

File: pkgs/build-support/compress-drv/default.nix [hunk 8]
```diff
@@ -70,9 +70,9 @@ let
     '';
   formatsPipe = builtins.concatStringsSep "|" formats;
 in
-runCommand "${drv.name}-compressed" { } ''
+runCommand "${drv.name}-compressed" { nativeBuildInputs = [ xorg.lndir ]; } ''
```
Comment (philiptaron) [inline pkgs/build-support/compress-drv/default.nix]: Curious: why do you prefer `nativeBuildInputs` above string interpolation?

File: pkgs/build-support/compress-drv/web.nix [hunk 9]
```diff
@@ -1,7 +1,8 @@
 {
-  zopfli,
   brotli,
   compressDrv,
+  lib,
```
Comment (philiptaron) [inline pkgs/build-support/compress-drv/web.nix:4]: Putting `lib` in the middle just looks super odd to me. Nothing wrong with it, I guess.

File: pkgs/build-support/compress-drv/default.nix (hunk 3 above)
Comment (philiptaron) [inline pkgs/build-support/compress-drv/default.nix:18]:
```suggestion
  : Extra command line parameters to pass to the [`find`](https://en.wikipedia.org/wiki/Find_(Unix)) command. This might be useful for excluding files. For example: `-not -iregex ".*(\/apps\/.*\/l10n\/).*"`
```

File: pkgs/build-support/compress-drv/web.nix [hunk 10]
```diff
@@ -132,8 +132,10 @@ drv:
     # https://github.com/NixOS/nixpkgs/pull/332752#issuecomment-2275110390
     zstd = "${lib.getExe zstd} --force --keep --quiet -19 {}";
   },
+  extraFindOperands ? "",
```
Comment (philiptaron) [inline pkgs/build-support/compress-drv/web.nix:135]: Please add this to the documentation of the Inputs section.

File: pkgs/build-support/compress-drv/default.nix (hunk 6 above)
Comment (philiptaron) [reply pkgs/build-support/compress-drv/default.nix:87]: There's an answer in the commit text of 5938ef4cff5a88997cd6d3421e35a3d14fed20e6:

> This is required to overwrite to use a compressed version of nextcloud in services.nextcloud.package because the module accesses version

File: pkgs/build-support/compress-drv/default.nix (hunk 5 above)
Comment (philiptaron) [reply pkgs/build-support/compress-drv/default.nix:91]: I think because the output doesn't really mean much.

**Review by philiptaron — COMMENTED** [review]
I'm only not approving because `nix-build -A gitea.data-compressed` fails for me. It also fails on master!

File: pkgs/by-name/gi/gitea/package.nix (hunk 7 above)
Comment (motiejus) [reply pkgs/by-name/gi/gitea/package.nix:91]: I'll have a look at this separately.

Comment (motiejus) [reply pkgs/by-name/gi/gitea/package.nix:91]: That's because of `lndir`. Now I realized why I didn't use it. I will update `compressDrv` soon to not use `lndir`.

The culprit is here: `-compressed` is the derivation being built. `public` should be a directory (so `compressDrv` can write files there), but is instead a symlink:

```
$ ls -lh /nix/store/yvsl8qm4y4c0xhfjq530ngnw1b54wj85-gitea-1.22.1-compressed/public
lrwxrwxrwx 1 nixbld1 nixbld 72 Aug 18 22:28 /nix/store/yvsl8qm4y4c0xhfjq530ngnw1b54wj85-gitea-1.22.1-compressed/public -> /nix/store/6yiik4x5r8midqp8shvc3pfvpac9bj6a-gitea-frontend-1.22.1/public
```

Comment (motiejus) [reply pkgs/by-name/gi/gitea/package.nix:91]: #335670

**Comment by philiptaron** [comment]
Needs a rebase now that https://github.com/NixOS/nixpkgs/pull/335670 is merged. @SuperSandro2000

File: pkgs/build-support/compress-drv/default.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:91]: Yeah, it just prints all the directories it creates which is really noisy.

File: pkgs/build-support/compress-drv/default.nix (hunk 8 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix]: It is cross friendlier and the run commands are shorter.

**Comment by SuperSandro2000** [comment]
> I don't understand the need for `extraFindOperands` given the current usecase.

see https://github.com/NixOS/nixpkgs/pull/332752#discussion_r1707153815

File: pkgs/build-support/compress-drv/web.nix (hunk 9 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/web.nix:4]: It's sorted alphabetically.

File: pkgs/build-support/compress-drv/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:18]: Not going to link wikipedia here, that won't help people.

File: pkgs/build-support/compress-drv/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:82]: That is used here:

https://github.com/NixOS/nixpkgs/blob/b46afe538da7262ce53b9f32b4a5c4d715d79801/nixos/modules/services/web-apps/nextcloud.nix#L850-L854

and I do the following in my config.

```
services.nextcloud.package = pkgs.compressDrvWeb pkgs.nextcloud29 {
  extraFindOperands = ''-not -iregex ".*(\/apps\/.*\/l10n\/).*"'';
};
```

File: pkgs/build-support/compress-drv/default.nix [hunk 11]
```diff
@@ -61,18 +71,22 @@ let
...
+    // (lib.optionalAttrs (drv ? version) { inherit (drv) version; })
+  )
+  ''
+    mkdir $out
 
-  # cannot use lndir here, because it also symlinks directories,
-  # which we do not need; we only need to symlink files.
-  (cd ${drv}; find -L -type d -exec mkdir -p $out/{} ';')
-  (cd ${drv}; find -L -type f -exec ln -s ${drv}/{} $out/{} ';')
+    # cannot use lndir here, because it stop recursing at symlinks that point to directories
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:87]: thanks

File: pkgs/build-support/compress-drv/default.nix [hunk 12]
```diff
@@ -61,18 +71,22 @@ let
     ext: prog:
     assert validProg ext prog;
     ''
-      find -L $out -type f -regextype posix-extended -iregex '.*\.(${formatsPipe})' -print0 \
+      find -L $out -type f -regextype posix-extended -iregex '.*\.(${formatsPipe})' ${extraFindOperands} -print0 \
         | xargs -0 -P$NIX_BUILD_CORES -I{} ${prog}
     '';
-  formatsPipe = builtins.concatStringsSep "|" formats;
+  formatsPipe = lib.concatStringsSep "|" formats;
```
Comment (motiejus) [inline pkgs/build-support/compress-drv/default.nix:77]: curious, why use lib over builtin?

**Review by motiejus — APPROVED** [review]

**Review by philiptaron — APPROVED** [review]
Built, tested, approved.

File: pkgs/build-support/compress-drv/default.nix (hunk 12 above)
Comment (SuperSandro2000) [reply pkgs/build-support/compress-drv/default.nix:77]: In this case it makes really no difference but some lib functions contain backwards compatibility code like https://github.com/NixOS/nixpkgs/blob/c3ede4ac04162b29c0fecc2208c9ceac44eec32e/lib/deprecated/misc.nix#L211-L213 and it is easier to just use lib for everything.
